examples/CustomCollectionView/Sample/MosaicCollectionLayoutDelegate.m

```diff
@@ -0,0 +1,168 @@
+  CGFloat top = 0;
+  
+  NSMapTable<ASCollectionElement *, UICollectionViewLayoutAttributes *> *attrsMap = [NSMapTable mapTableWithKeyOptions:(NSMapTableObjectPointerPersonality | NSMapTableWeakMemory) valueOptions:NSMapTableStrongMemory];
+  NSMutableArray *columnHeights = [NSMutableArray array];
+  
+  NSInteger numberOfSections = [elements numberOfSections];
+  for (NSUInteger section = 0; section < numberOfSections; section++) {
+    NSInteger numberOfItems = [elements numberOfItemsInSection:section];
+    
+    top += _sectionInset.top;
+    
+    if (_headerHeight > 0) {
+      NSIndexPath *indexPath = [NSIndexPath indexPathForItem:0 inSection:section];
+      ASCollectionElement *element = [elements supplementaryElementOfKind:UICollectionElementKindSectionHeader
+                                                              atIndexPath:indexPath];
+      UICollectionViewLayoutAttributes *attrs = [UICollectionViewLayoutAttributes layoutAttributesForSupplementaryViewOfKind:UICollectionElementKindSectionHeader
+                                                                                                                     withIndexPath:indexPath];
+      
+      ASSizeRange sizeRange = [self sizeRangeForHeaderOfSection:section withLayoutWidth:layoutWidth];
+      CGSize size = [element.node layoutThatFits:sizeRange].size;
+      CGRect frame = CGRectMake(_sectionInset.left, top, size.width, size.height);
+      
+      attrs.frame = frame;
+      [attrsMap setObject:attrs forKey:element];
+      top = CGRectGetMaxY(frame);
+    }
+    
+    [columnHeights addObject:[NSMutableArray array]];
+    for (NSUInteger idx = 0; idx < _numberOfColumns; idx++) {
+      [columnHeights[section] addObject:@(top)];
+    }
+    
+    CGFloat columnWidth = [self _columnWidthForSection:section withLayoutWidth:layoutWidth];
+    for (NSUInteger idx = 0; idx < numberOfItems; idx++) {
+      NSUInteger columnIndex = [self _shortestColumnIndexInSection:section withColumnHeights:columnHeights];
+      NSIndexPath *indexPath = [NSIndexPath indexPathForItem:idx inSection:section];
+      ASCollectionElement *element = [elements elementForItemAtIndexPath:indexPath];
+      UICollectionViewLayoutAttributes *attrs = [UICollectionViewLayoutAttributes layoutAttributesForCellWithIndexPath:indexPath];
+      
+      ASSizeRange sizeRange = [self sizeRangeForItem:element.node atIndexPath:indexPath withLayoutWidth:layoutWidth];
+      CGSize size = [element.node layoutThatFits:sizeRange].size;
+      CGPoint position = CGPointMake(_sectionInset.left + (columnWidth + _columnSpacing) * columnIndex,
+                                         [columnHeights[section][columnIndex] floatValue]);
+      CGRect frame = CGRectMake(position.x, position.y, size.width, size.height);
+      
+      attrs.frame = frame;
+      [attrsMap setObject:attrs forKey:element];
+      columnHeights[section][columnIndex] = @(CGRectGetMaxY(frame) + _interItemSpacing.bottom);
```

Comment:
Vector might be worthwhile to avoid boxing - try profiling this first and look for object retain / release overhead. Apps will probably borrow this sample code and ship it :)